foundation-ui/navigation/NavigationService.js

```diff
@@ -0,0 +1,111 @@
+import {EventEmitter} from 'events';
+import {NAVIGATION_CHANGE} from './EventTypes';
+
+class NavigationService extends EventEmitter {
+
+  constructor() {
+    super();
+
+    this.deferredTasks = [];
+
+    // The only default is root, so it goes first no matter what
+    this.definition = [
+      { category: 'root', children: [] }
+    ];
+
+    this.on(
+      NAVIGATION_CHANGE,
+      this.processDeferred
+    );
+  }
+
+  getDefinition() {
+    return this.definition.slice(0);
+  }
+
+  processDeferred() {
+    const tasks = this.deferredTasks.slice(0);
+    this.deferredTasks = [];
+
+    // If Task is unable to resolve at this point of time it will re-defer itself
+    tasks.forEach(({method, args}) => {
+      this[method].apply(this, args);
+    });
+  }
+
+  defer(method, args) {
+    this.deferredTasks.push({method, args});
+  }
+
+  registerCategory(category) {
+    if (!this.definition.find((element) => element.category === category)) {
+      this.definition.push({
+        category,
+        children: []
+      });
+    }
+
+    this.emit(NAVIGATION_CHANGE);
+  }
+
+  registerPrimary(path, link, options = {}) {
```

Comment:
In my taste it would be again a function with 4 arguments and declaration would be rather hard to read. Right now it's straight forward what you want to register, category, primary, secondary